legwork/psd.py

```diff
@@ -0,0 +1,252 @@
+    def Poms(f):
+        return (1.5e-11)**2 * (1 + (2e-3 / f)**4)
+
+    # single test mass acceleration noise (Robson+ Eq. 11)
+    def Pacc(f):
+        return (3e-15)**2 * (1 + (0.4e-3 / f)**2) * (1 + (f / (8e-3))**4)
+
+    # galactic confusion noise (Robson+ Eq. 14)
+    def Sc(f, t_obs):
+
+        # parameters from Robson+ 2019 Table 1
+        lengths = np.array([0.5, 1.0, 2.0, 4.0]) * u.yr
+        alpha = [0.133, 0.171, 0.165, 0.138]
+        beta = [243.0, 292.0, 299.0, -221.0]
+        kappa = [482.0, 1020.0, 611.0, 521.0]
+        gamma = [917.0, 1680.0, 1340.0, 1680.0]
+        fk = [2.58e-3, 2.15e-3, 1.73e-3, 1.13e-3]
+
+        # find index of the closest length to inputted observation time
+        ind = np.abs(t_obs - lengths).argmin()
+
+        return 9e-45 * f**(-7/3.) \
+            * np.exp(-f**(alpha[ind]) + beta[ind] * f
+                     * np.sin(kappa[ind] * f)) \
+            * (1 + np.tanh(gamma[ind] * (fk[ind] - f)))
+
+    # calculate response function (either exactly or with approximation)
+    fstar = (const.c / (2 * np.pi * L * u.m)).to(u.Hz).value
+    if approximate_R:
+        R = approximate_response_function(f, fstar)
+    else:
+        R = load_response_function(f, fstar)
+
+    # work out the confusion noise or just set to 0
+    if include_confusion_noise:
+        cn = Sc(f, t_obs)
+    else:
+        cn = np.zeros(len(f)) if isinstance(f, (list, np.ndarray)) else 0.0
+
+    # calculate sensitivity curve
+    psd = (1 / (L**2) * (Poms(f) + 4 * Pacc(f) / (2 * np.pi * f)**4)) / R + cn
+
+    # replace values for bad frequencies (set to extremely high value)
+    psd = np.where(np.logical_and(f >= MIN_F, f <= MAX_F), psd, HUGE_NOISE)
+    return psd / u.Hz
+
+
+def tianqin_psd(f, L=np.sqrt(3) * 1e5 * u.km, t_obs=None, approximate_R=None,
+                include_confusion_noise=None):
+    """Calculates the effective TianQin power spectral density sensitivity
+    curve
+
+    Using equations from Huang+20, calculate the effective TianQin power
+    spectral density for the sensitivity curve
+
+    Parameters
+    ----------
+    f : `float/array`
+        Frequencies at which to evaluate the sensitivity curve
+
+    L : `float`
+        Arm length
+
+    t_obs : `float`
+        Observation time (ignored)
+
+    approximate_R : `boolean`
+        Whether to approximate the response function (ignored)
+
+    include_confusion_noise  : `boolean`
+        Whether to include the Galactic confusion noise (ignored)
+
+    Returns
+    -------
+    psd : `float/array`
+        Effective power strain spectral density
+    """
+    fstar = const.c / (2 * np.pi * L)
+    Sa = 1e-30 * u.m**2 * u.s**(-4) * u.Hz**(-1)
+    Sx = 1e-24 * u.m**2 * u.Hz**(-1)
+    psd = 1 / L**2 * (4 * Sa / (2 * np.pi * f)**4
+                      * (1 + (1e-4 * u.Hz / f)) + Sx) \
+        * (1 + 0.6 * (f / fstar)**2)
+    return psd.to(u.Hz**(-1))
+
+
+def power_spectral_density(f, instrument="LISA", custom_function=None,
```

Comment:
awesome move here! I was going to suggest that we do this, but you've already implemented it! 🥳

Reply:
As ever we are on the same wavelength!